org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/completion/DOMCompletionEngine.java

```diff
@@ -197,20 +287,84 @@ private Collection<? extends ITypeBinding> visibleTypeBindings(ASTNode node) {
 		}
 		return visibleBindings;
 	}
+	
+	private static CompilationUnit getCompletionAST(ITypeRoot typeRoot, IJavaProject javaProject,
+			WorkingCopyOwner workingCopyOwner, int focalPosition) {
+		Map<String, String> options = javaProject.getOptions(true);
```

Comment:
using `ICompilationUnit.getOrBuildAST(..., focalPosition)` here may allow to save a DOM creation: if a full AST was already created for hover, instead of rebuilding it here, we can just reuse it.